File: diploma-data/src/main/java/org/arhor/diploma/data/domain/DomainObject.java
package org.arhor.diploma.data.domain;

import lombok.Data;

import javax.persistence.Column;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.MappedSuperclass;

@Data
@MappedSuperclass
public abstract class DomainObject<T> implements Identifiable<T>, Deletable {

  @Id
  @GeneratedValue(strategy = GenerationType.IDENTITY)
  protected T id;

  @Column(nullable = false)
  private boolean deleted;
}

File: diploma-server/src/main/kotlin/org/arhor/diploma/domain/DomainObject.kt
package org.arhor.diploma.domain

import java.io.Serializable
import javax.persistence.*

@MappedSuperclass
abstract class DomainObject<T : Serializable> : Identifiable<T>, Deletable {

  @Id
  @GeneratedValue(strategy = GenerationType.IDENTITY)
  private var id: T? = null

  @Column(name = "deleted", nullable = false)
  private var isDeleted: Boolean = false

  override fun getId(): T? {
    return id
  }

  override fun setId(id: T?) {
    this.id = id
  }

  override fun isDeleted(): Boolean {
    return isDeleted
  }

  override fun setDeleted(isDeleted: Boolean) {
    this.isDeleted = isDeleted
  }

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (other !is DomainObject<*>) return false

    if (id != other.id) return false
    if (isDeleted != other.isDeleted) return false

    return true
  }

  override fun hashCode(): Int {
    var result = id?.hashCode() ?: 0
    result = 31 * result + isDeleted.hashCode()
    return result
  }
}
